File: python/Best Time to Buy and Sell Stock.py
class Solution(object):
    def maxProfit(self, prices):
        if len(prices) <= 1:
            return 0

        answer = prices[1] - prices[0]
        if answer < 0: answer = 0
        dp = [0] * len(prices)
        
        for i in range(1, len(prices)):
            dp[i] = max(dp[i - 1] + prices[i] - prices[i - 1], prices[i] - prices[i - 1])
            answer = max(answer, dp[i])
        
        return answer

File: cpp/Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() <= 1) {
            return 0;
        }
        int minBuy = prices[0];
        int answer = 0;
        for (int i = 0; i < prices.size(); i++) {
            if (prices[i] < minBuy) {
                minBuy = prices[i];
            }
            if (prices[i] - minBuy > answer) {
                answer = prices[i] - minBuy;
            }
        }
        return answer;
    }
};
